tests/manifold/tensor_product_manifold_03.cc

```diff
@@ -0,0 +1,52 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 - 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Test the correct handling of data members within the TensorProductManifold
+// class. The test generates a general hypercube grid and, in an inner scope,
+// attempts to assign a TensorProductManifold to it. The test is succesfull if
+// it terminates without subscriptor errors or memory leaks.
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/manifold_lib.h>
+#include <deal.II/grid/tensor_product_manifold.h>
+
+#include "../tests.h"
+
+
+void
+test()
+{
+  Triangulation<3, 3> tria;
+  GridGenerator::hyper_cube(tria);
+  {
+    FunctionManifold<1, 1>                     F("x", "x");
+    PolarManifold<2, 2>                        G(Point<2>(0.5, 0.5));
+    TensorProductManifold<3, 2, 2, 2, 1, 1, 1> manifold(G, F);
+    tria.set_all_manifold_ids(0);
+    tria.set_manifold(0, manifold);
+  }
```

Comment:
was the original problem that the `manifold` variable goes out of scope here but the `tria` object still keeps a pointer to it?

Reply:
You get
```
141: manifold/tensor_product_manifold_03.debug: RUN failed. ------ Return code 134
141: manifold/tensor_product_manifold_03.debug: RUN failed. ------ Result: /mnt/data/darndt/Sources/dealii-clang-6.0.0/build/tests/manifold/tensor_product_manifold_03.debug/failing_output
141: manifold/tensor_product_manifold_03.debug: RUN failed. ------ Partial output:
141: 
141: --------------------------------------------------------
141: An error occurred in line <108> of file <../source/base/subscriptor.cc> in function
141:     void dealii::Subscriptor::check_no_subscribers() const
141: The violated condition was: 
141:     counter == 0
141: Additional information: 
141:     Object of class N6dealii13PolarManifoldILi2ELi2EEE is still used by 1 other objects.
141: 
141: (Additional information: 
141:   from Subscriber N6dealii21TensorProductManifoldILi3ELi2ELi2ELi2ELi1ELi1ELi1EEE)
141: 
141: See the entry in the Frequently Asked Questions of deal.II (linked to from http://www.dealii.org/) for a lot more information on what this error means and how to fix programs in which it happens.
141: --------------------------------------------------------
141: 
141: Aborted
```

Reply:
No, actually not the Tria object, but the TensorManifold inside the Tria object. It is the Manifold who is storing references to the base manifolds